packages/material-ui-unstyled/src/SwitchUnstyled/SwitchUnstyled.tsx

```diff
@@ -0,0 +1,211 @@
+import * as React from 'react';
+import PropTypes from 'prop-types';
+import clsx from 'clsx';
+import useSwitch, { SwitchState, UseSwitchProps } from './useSwitch';
+import classes from './switchUnstyledClasses';
+import { getForwardableProps, isHostComponent } from '../utils';
+
+export interface SwitchUnstyledProps extends UseSwitchProps {
+  /**
+   * Class name applied to the root element.
+   */
+  className?: string;
+  /**
+   * The component used for the Root slot.
+   * Either a string to use a HTML element or a component.
+   * This is equivalent to `components.Root`. If both are provided, the `component` is used.
+   */
+  component?: React.ElementType;
+  /**
+   * The components used for each slot inside the Switch.
+   * Either a string to use a HTML element or a component.
+   * @default {}
+   */
+  components?: {
+    Root?: React.ElementType;
+    Thumb?: React.ElementType;
+    Input?: React.ElementType;
+  };
+
+  /**
+   * The props used for each slot inside the Switch.
+   * @default {}
+   */
+  componentsProps?: {
+    root?: {};
+    thumb?: {};
+    input?: {};
+  };
+}
+
+const appendStyleProps = (
+  component: React.ElementType,
+  componentsProps: Record<string, any>,
+  state: SwitchState,
+) => {
+  if (!isHostComponent(component)) {
+    componentsProps.styleProps = { ...componentsProps.styleProps, ...state };
+  }
+};
+
+/**
+ * The foundation for building custom-styled switches.
+ *
+ * Demos:
+ *
+ * - [Switches](https://material-ui.com/components/switches/)
+ *
+ * API:
+ *
+ * - [SwitchUnstyled API](https://material-ui.com/api/switch-unstyled/)
+ */
+const SwitchUnstyled = React.forwardRef(function SwitchUnstyled(
+  props: SwitchUnstyledProps,
+  ref: React.ForwardedRef<any>,
+) {
+  const {
+    checked: checkedProp,
+    className,
+    component,
+    components = {},
+    componentsProps = {},
+    defaultChecked,
+    disabled: disabledProp,
+    onBlur,
+    onChange,
+    onFocus,
+    onFocusVisible,
+    readOnly: readOnlyProp,
+    required,
+    ...otherProps
+  } = props;
+
+  const Root: React.ElementType = component ?? components.Root ?? 'span';
+  const rootProps: any = { ...getForwardableProps(Root, otherProps), ...componentsProps.root };
```

Comment:
Why is this not enough?

```suggestion
  const rootProps: any = { ...otherProps, ...componentsProps.root };
```

Reply:
If `Root` is a plain HTML element, I don't want to forward anything that doesn't apply to a HTML element.

Reply:
The developers can swallow the extra prop. We only check it with the "other" props on the root element, but not with the `componentsProps` slots.

I mean, I don't get why foo shouldn't warn when it does with the buttons:

```jsx
import * as React from "react";
import Button from "@material-ui/core/Button";
import SwitchUnstyled from "@material-ui/unstyled/SwitchUnstyled";

export default function UnstyledSwitches() {
  return (
    <div>
      <SwitchUnstyled foo />
      <Button foo />
      <button foo />
    </div>
  );
}
```

https://codesandbox.io/s/unstyledswitches-material-demo-forked-rsrd1?file=/demo.tsx

Reply:
My thinking was that forwarding props from the host to the root is less explicit than just setting them in `componentsProps.root` and therefore it may need some additional checking. If someone explicitly passes something in `componentsProps.*`, I assume they know what they are doing.

This has an inconvenience of being inconsistent, though. I'm not particularly attached to this piece of code and if you feel we should always forward everything, so be it.

Reply:
cc @mui-org/core for more perspectives on the matter.

Outside of this point the PR is good to go, from my side.

Reply:
I would forward all props, as it was done before. Let's not complicate the code for problems we haven't encounter so far.